cpp/arcticdb/column_store/column_data.hpp

```diff
@@ -133,6 +133,78 @@ struct ColumnData {
  */
 
   public:
+    template<typename TDT, bool constant>
+    class ColumnDataIterator: public boost::iterator_facade<
+            ColumnDataIterator<TDT, constant>,
+            std::conditional_t<constant, typename TDT::DataTypeTag::raw_type const, typename TDT::DataTypeTag::raw_type>,
+            boost::forward_traversal_tag
+            > {
+    using RawType = std::conditional_t<constant, const typename TDT::DataTypeTag::raw_type, typename TDT::DataTypeTag::raw_type>;
+    public:
+        ColumnDataIterator() = delete;
+
+        // Used to construct [c]begin iterators
+        explicit ColumnDataIterator(ColumnData* parent):
+        parent_(parent)
+        {
+            opt_block_ = parent->next<TDT>();
+            if(opt_block_.has_value()) {
+                remaining_values_in_block_ = opt_block_->row_count();
+                if constexpr(constant) {
+                    ptr_ = reinterpret_cast<RawType*>(opt_block_->data());
+                } else {
+                    ptr_ = const_cast<RawType*>(opt_block_->data());
+                }
+            }
+        }
+
+        // Used to construct [c]end iterators
+        explicit ColumnDataIterator(ColumnData* parent, RawType* end_ptr):
+                parent_(parent),
+                ptr_(end_ptr) {}
+
+        template <class OtherValue, bool OtherConst>
+        ColumnDataIterator(ColumnDataIterator<OtherValue, OtherConst> const& other):
+        parent_(other.parent_),
+        opt_block_(other.opt_block_),
+        remaining_values_in_block_(other.remaining_values_in_block_),
+        ptr_(other.ptr_) {}
+    private:
+        friend class boost::iterator_core_access;
+        template <class, bool> friend class ColumnDataIterator;
+
+        void increment() {
+            if (ARCTICDB_LIKELY(remaining_values_in_block_ > 0)) {
+                ++ptr_;
+                --remaining_values_in_block_;
+            } else {
+                opt_block_ = parent_->next<TDT>();
+                if(ARCTICDB_LIKELY(opt_block_.has_value())) {
+                    remaining_values_in_block_ = opt_block_->row_count();
+                    if constexpr(constant) {
+                        ptr_ = reinterpret_cast<RawType*>(opt_block_->data());
+                    } else {
+                        ptr_ = const_cast<RawType*>(opt_block_->data());
+                    }
+                }
+            }
+        }
+
+        template <typename OtherValue, bool OtherConst>
+        bool equal(ColumnDataIterator<OtherValue, OtherConst> const& other) const {
+            return parent_ == other.parent_ && ptr_ == other.ptr_;
+        }
+
+        RawType& dereference() const {
+            return *ptr_;
+        }
+
+        ColumnData* parent_{nullptr};
+        std::optional<TypedBlockData<TDT>> opt_block_{std::nullopt};
+        std::size_t remaining_values_in_block_{0};
+        RawType* ptr_{nullptr};
```

Comment:
For the sparse case, we'll still need a `RawType*` for retrieving the value. I think a `row_idx_` (or similar) member will be needed for the `enumerate `style methods, and that will be a `size_t` for dense columns, and a bit magic enumerator for sparse columns (as you did in the aggregators). The idea will be to have the `if` as high up as possible, so that it is only branched on once, rather than for every value.